Feature/links with city objects

Crejak opened the pull request:
# Changes

Completed scenario 2 for the links

- Added a `CityObject` module to explore and filter city objects
  - Ability to "select" a city object with the mouse to see its info (similar to 3DTiles debug)
  - Possibility to change the style (color) of city objects according to filters (default filter = based on the ID)
- The `Link` module now serves as a binding between the city object and the document module.
  - It adds extensions to both modules to display the correct information
  - It manages the interaction between the modules : when a city object is selected, or when the displayed document change, the link module is able to update the view accordingly

Other changes

- Moved the "extensions" system in the `Window` class, which means any window can accept extensions. An example is provided in the `DemoWindow` demo.
- Added a `findAllCityObjects` method in `TilesManager` to filter city objects according to a given predicate

# Screenshot

![image](https://user-images.githubusercontent.com/7014405/63516174-c81a3f00-c4ec-11e9-9060-263ecef7938b.png)

jailln commented on UDV-Core/examples/DemoWindow/Demo.js:
```diff
@@ -24,6 +24,32 @@ function main () {
+        html: '[bouton]',
+        container: 'bottom',
+        callback: () => console.log('test bouton')
+    })
```
English please :)

Crejak replied:
Changed

jailln commented on UDV-Core/src/Extensions/Contribute/View/DocumentCreationWindow.js:
```diff
@@ -147,6 +157,22 @@ export class DocumentCreationWindow extends AbstractDocumentWindow {
+    this.controls.maxPanSpeed = 15.0;
+    this.controls.minPanSpeed = 0.05;
+  }
+
```
Could you comment these functions please ?

Crejak replied:
Done (also a bit refactored)

jailln commented on UDV-Core/src/Modules/Links/Model/Link.js:
```diff
@@ -0,0 +1,52 @@
+     * @type {number}
+     */
+    this.centroid_z;
+  }
```
Is this centroïd needed to travel to the cityobject ? If yes, could you add a comment about that please ?

Crejak replied:
Yes, I updated the comments to explain

jailln commented on UDV-Core/src/Utils/BaseDemo/js/BaseDemo.js:
```diff
@@ -482,7 +482,7 @@ export class BaseDemo {
 
         // Controls
-        this.controls = new itowns.PlanarControls(this.view, {});
+        this.controls = new itowns.PlanarControls(this.view, {maxZenithAngle: 180, groundLevel: -100, handleCollision: false});
```
Maybe this should be moved to `DocumentCreationWindow._enterEditMode()` ?

Crejak replied:
I don't think so, because the positioning and the visualization of the documents should have the same parameters (at least of these three options). Otherwise, we could have situations where a document is placed somewhere the camera can't reach with the default settings (too close to the ground, zenith angle out of bounds, etc.), and in this case the visualization is broken.

jailln replied:
Yep you're right, my bad.

Crejak commented:
I added a few commits to fix the PR.

- Concerning the ergonomy, I made the changes in the user interface (added a clear button), and handled the case where the city object window is closed (styles are cleared).
- For the name 'Layer', I couldn't come up with a simple and understandable name. For the moment I didn't change it, I think we can keep it until we find a new name

I also updated some docs, and added a new 'stable' demo along with the full modules demo.

jailln commented on README.md:
```diff
@@ -7,7 +7,7 @@ Server-side tools can be found [here](https://github.com/MEPP-team/UDV-server).
 ## Demo
 Online demos (alas only visible on the [Lyon1](https://sciences.univ-lyon1.fr/)/[INSA-Lyon](https://www.insa-lyon.fr/en/) campus):
- - [UDV](http://rict.liris.cnrs.fr/UDVDemo-2/UDV/UDV-Core/)
+ - [UDV](http://rict.liris.cnrs.fr/UDV/UDV-Core/)
```
Broken link and I think that the demos are visible outside from the campus now

Crejak replied:
Yes, I fixed it and removed the "only visible on..."

jailln commented on UDV-Core/examples/data/config/generalDemoConfig.json:
```diff
@@ -1,7 +1,8 @@
   "type": "class",
   "server":{
-    "url":"http://rict.liris.cnrs.fr:443/",
+    "url":"http://rict.liris.cnrs.fr:1525/",
```
devUrl ?

jailln commented:
Ok to keep the name Layer for now as it is documented in the code.
